qsv: hevc10 encode support

maximd33 opened the pull request:
pretty much simple implementation changes needed for support of HEVC 10b support with QSV while on latest Kaby Lake HW.

libav patch adds support of P010 surfaces for swscale on output,
merge-commit from ffmpeg.

PS: Pull request includes empty commits(like upstream merging...) but looks that github engine should be blamed(?)

maximd33 committed:
qsv: adding hevc10 support starting from KBL platform

maximd33 committed:
qsv: libav's patch fix

bradleysepos commented:
I can help resolve the merge commits issue at some point. Just needs a manual rebase and conflict resolution. Or, you can create a new branch from latest master and cherry-pick the relevant commits.

maximd33 commented:
thank you

>you can create a new branch from latest master and cherry-pick the relevant commits.

I should try it next pull request,
assuming it will not include again merging of upstream/master

sr55 commented:
From a front-end presepective, I can make any changes if necessary, but I think I updated the profiles to be pulled from libhb, so it should mostly be ok.

Any word on when libav will commit pathces upstream?

maximd33 commented:
sr55, as I understood - it is hard to predict about libav
due to its complex situation with swscale support in general.

sr55 commented:
Fair enough.

sr55 commented:
This will need resolved.  I can't actually test it. Maybe @Rodeo314  can but I can certainly check we've not regressed anything.

sr55 commented on libhb/qsv_common.c:
```diff
@@ -1542,6 +1542,15 @@ int hb_qsv_profile_parse(hb_qsv_param_t *param, hb_qsv_info_t *info, const char
+                if (profile->value == MFX_PROFILE_HEVC_MAIN10 &&
+                    qsv_hardware_generation(hb_get_cpu_platform()) < QSV_G6)
+                {
+                    hb_log("HEVC Main10 is not supported on this platform");
```
Maybe add qsv:  at the start of the message for clarity.

sr55 replied:
I wonder if we also need to update hb_video_encoder_get_profiles. The WinGUI uses this for getting profiles. (oddly I don't see QSV in there so I wonder how that even works)

Point being, we could make sure main10 is not available in the gui on pre kaby hardware.

Thoughts?

maximd33 replied:
good points: will fix first one and second: should yet try GUI (prefer CLI therefore these changes)

@sr55 would you have any time to try on KBL HW ? 
if so , I will think to share access to my test PC.

sr55 replied:
My newest hardware is skylake only but if you can provide remote access I can give it a quick sanity check if you wish.  I can confirm no regression on my haswell/skylake hardware

maximd33 committed:
qsv: added HEVC 10-bit as encoder, should help for GUI integration

maximd33 committed:
qsv: only main10 profile supported for HEVC 10b encode

bradleysepos commented:
@maximd33 Here are instructions to fix this branch from the command line. We will rewind the branch to the commit before your first change, rebase the stale but unaltered branch on current master (fast-forwarding to now), and then apply your four commits (omitting 7656a77 since it was previously committed to master).

```
git checkout hevc10_enc
git reset --hard 7656a77~1
git rebase master
git cherry-pick dece2b1
git cherry-pick 145e006
git cherry-pick 0ca416
git cherry-pick 5882fec
```

After this, you may update the pull request by force pushing your branch to GitHub, e.g. `git push --force <remote> hevc10_enc`.

bradleysepos reviewed (changes requested):
See my previous comment regarding git, then I approve.

maximd33 commented:
@bradleysepos good idea but it looks that something missing yet, see now:
- new commit 963a5c39ab5dd6198a460529655bd799e1753576 shows in commits list ?

ideas to fix this ?

bradleysepos commented:
Hmm, it would appear perhaps that you didn't rewind entirely first, or perhaps there was some conflict I didn't experience. Assuming 963a5c3 is unwanted, you should be able to follow my directions again. Be sure to update master first (`git checkout master && git pull --ff-only <remote> master`).

```
$ git checkout hevc10_enc
Switched to branch 'hevc10_enc'
$ git reset --hard 7656a77~1
HEAD is now at 39d03522f WinGui: Show Queue button can now maximise a minimised window. Fixes #262
$ git rebase master
First, rewinding head to replay your work on top of it...
Fast-forwarded hevc10_enc to master.
$ git cherry-pick dece2b1
[hevc10_enc 040d88d3c] qsv: adding hevc10 support starting from KBL platform
 Author: maxd <maxim.d33@gmail.com>
 Date: Wed Feb 8 13:17:00 2017 +0100
 3 files changed, 203 insertions(+), 14 deletions(-)
 create mode 100644 contrib/ffmpeg/A05-p10-output-support.patch
$ git cherry-pick 145e006
[hevc10_enc 4414b835f] qsv: libav's patch fix
 Author: maxd <maxim.d33@gmail.com>
 Date: Wed Feb 8 13:39:21 2017 +0100
 1 file changed, 9 insertions(+), 9 deletions(-)
$ git cherry-pick 0ca416
[hevc10_enc 0421fdd26] qsv: added HEVC 10-bit as encoder, should help for GUI integration
 Author: maxd <maxim.d33@gmail.com>
 Date: Tue Feb 28 18:15:00 2017 +0100
 5 files changed, 27 insertions(+), 6 deletions(-)
$ git cherry-pick 5882fec
[hevc10_enc 798adcb36] qsv: only main10 profile supported for HEVC 10b encode
 Author: maxd <maxim.d33@gmail.com>
 Date: Tue Feb 28 18:27:59 2017 +0100
 2 files changed, 9 insertions(+), 6 deletions(-)
$ git shortlog -5
maxd (4):
      qsv: adding hevc10 support starting from KBL platform
      qsv: libav's patch fix
      qsv: added HEVC 10-bit as encoder, should help for GUI integration
      qsv: only main10 profile supported for HEVC 10b encode

sr55 (1):
      WinGui: Rework of the preset system - Better support for categories. - Switched from a List to Treeview Control. - Remember the expansion state of each group - Put User Presets on top after next save. Closes #445

```

maximd33 commented:
> mdmytryc-mac01:HandBrake mdmytryc$ git checkout master && git pull --ff-only master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
fatal: 'master' does not appear to be a git repository
fatal: Could not read from remote repository.
Please make sure you have the correct access rights
and the repository exists.
mdmytryc-mac01:HandBrake mdmytryc$ git checkout master && git pull --ff-only origin
Already on 'master'
Your branch is up-to-date with 'origin/master'.
remote: Counting objects: 10, done.
remote: Total 10 (delta 9), reused 9 (delta 9), pack-reused 1
Unpacking objects: 100% (10/10), done.
From https://github.com/maximd33/HandBrake
   5c8e17a..64c6f07  hevc10_enc -> origin/hevc10_enc
Already up-to-date.
mdmytryc-mac01:HandBrake mdmytryc$  git checkout hevc10_enc
Switched to branch 'hevc10_enc'
Your branch is behind 'origin/hevc10_enc' by 149 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
mdmytryc-mac01:HandBrake mdmytryc$ git reset --hard 7656a77~1
HEAD is now at 39d0352 WinGui: Show Queue button can now maximise a minimised window. Fixes #262
mdmytryc-mac01:HandBrake mdmytryc$ git cherry-pick dece2b1
error: could not apply dece2b1... qsv: adding hevc10 support starting from KBL platform
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'

why I have the conflict but only me?

BTW, I am fine if you can apply patches and commit...
....these pull request problems ......

bradleysepos commented:
> why I have the conflict but only me?

You forgot `git rebase master` before cherry picking.

bradleysepos commented:
Since you enabled maintainer edits when you created the pull request, I was allowed to push to your GitHub branch. Should be good now.
